[clang-format] Add BreakBeforeTemplateCloser option

leijurv opened the pull request:
In clang-format, multiline templates have the `>` on the same line as the last parameter:

```c++
template <
    typename Foo,
    typename Bar>
void foo() {
```

I would like to add an option to put the `>` on the next line, like this:

```c++
template <
    typename Foo,
    typename Bar
>
void foo() {
```

An example of a large project that uses this style is NVIDIA's CUTLASS, here is an example:

https://github.com/NVIDIA/cutlass/blob/main/include/cutlass/epilogue/dispatch_policy.hpp#L149-L156

My reasoning is that it reminds me of this style of braces:

```c++
if (foo()) {
    bar();
    baz();}
```

Most people agree this is better:

```c++
if (foo()) {
    bar();
    baz();
}
```

leijurv committed:
[clang-format] Add BreakBeforeTemplateClose option

leijurv commented:
Ping :pray:

gedare commented:
Would this also be equivalent to supporting `AlignAfterOpenBracket: BlockIndent` for angle brackets?

leijurv commented:
Yes. I followed the code that implements `AlignAfterOpenBracket: BlockIndent`, except applying to ClosesTemplateDeclaration. I made it a new setting for backwards compatibility. 👍

gedare commented:
> Yes. I followed the code that implements `AlignAfterOpenBracket: BlockIndent`, except applying to ClosesTemplateDeclaration. I made it a new setting for backwards compatibility. 👍

Maybe we should make it a sub-option for `BlockIndent`. I'm seeing a possible need for having control over other kinds of block indented structures such as Lambdas. I suspect the way this option has been implemented it interacts with the alignment options anyway. This would be a question probably for @owenca

gedare commented:
> > Yes. I followed the code that implements `AlignAfterOpenBracket: BlockIndent`, except applying to ClosesTemplateDeclaration. I made it a new setting for backwards compatibility. 👍
> 
> Maybe we should make it a sub-option for `BlockIndent`. I'm seeing a possible need for having control over other kinds of block indented structures such as Lambdas. I suspect the way this option has been implemented it interacts with the alignment options anyway. This would be a question probably for @owenca

I thought about it some more and I don't think it would be a good idea. Having a separate option seems a bit easier to handle, and allows mixing and matching by users.

gedare commented:
I think you should add test cases with multiple `>>` closing, see `UnderstandsTemplateParameters` for example.

Consider trying the code snippet mentioned in https://github.com/llvm/llvm-project/issues/80049 as well.

leijurv committed:
add test case

leijurv commented:
> I think you should add test cases with multiple `>>` closing, see `UnderstandsTemplateParameters` for example.
> 
> Consider trying the code snippet mentioned in #80049 as well.

Thank you for the suggestion! I have integrated that test case. It indicated that I was being too narrow - `ClosesTemplateDeclaration` only applies before functions, but, as you demonstrate, there are several other cases where templates are closed, so I have switched to `TT_TemplateCloser`.

https://github.com/llvm/llvm-project/pull/118046/commits/212e444426a2d194df996909a40ad0e0fc681d98

HazardyKnusperkeks commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -11077,6 +11077,157 @@ TEST_F(FormatTest, WrapsTemplateDeclarationsWithComments) {
       Style);
 }
 
+TEST_F(FormatTest, BreakBeforeTemplateClose) {
```
I want to see some lambdas and template usages (not declarations).

leijurv replied:
Good suggestion! I have added tests in this commit: https://github.com/llvm/llvm-project/pull/118046/commits/3927d411eeece27521e1a4034d01012c84beaf4b

leijurv committed:
more tests

HazardyKnusperkeks commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -11077,6 +11077,221 @@ TEST_F(FormatTest, WrapsTemplateDeclarationsWithComments) {
+      "}\n",
+      Style);
+
+  // test lambda goes to next line:
```
Can you also add a lambda where the template is not on a new line? And multiple template parameters?

leijurv replied:
Sure! https://github.com/llvm/llvm-project/pull/118046/commits/2c3a64c8f32a5c43887db4d107a4142490898c20

HazardyKnusperkeks commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -11077,6 +11077,221 @@ TEST_F(FormatTest, WrapsTemplateDeclarationsWithComments) {
       Style);
 }
 
+TEST_F(FormatTest, BreakBeforeTemplateClose) {
```
One additional test: How about nested templates?
`template<template<typename, typename> typename T>` for example.

leijurv replied:
Sure! https://github.com/llvm/llvm-project/pull/118046/commits/2c3a64c8f32a5c43887db4d107a4142490898c20

HazardyKnusperkeks replied:
This one is still missing.

leijurv replied:
Sorry if I misunderstood, but, I thought I added that here: 
![Screenshot_20241224-154618](https://github.com/user-attachments/assets/ce0a3af9-421e-48a9-a8ec-9b443f26b324)

HazardyKnusperkeks replied:
Sorry missed that, didn't look for the lambda case.

leijurv committed:
more tests

HazardyKnusperkeks commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -11077,6 +11077,281 @@ TEST_F(FormatTest, WrapsTemplateDeclarationsWithComments) {
+      // nested with long capture:
+      "  auto lambda =\n"
+      "      [loooooooooooooooooooong]<\n"
+      "          template <typename, typename>\n"
```
Now the question is, shouldn't this also break?
```suggestion
      "          template <typename, typename\n"
      "          >\n"
```

leijurv replied:
No, I don't think so, because the matching `<` is on the same line. In this part of the test, there is a column limit, and it pulls the template back onto one line whenever it can, which I think is good behavior. When there is no column limit, it keeps the state of multiple lines versus one line, which is also covered by the tests.

The analogy would be: if you set `AlignAfterOpenBracket: BlockIndent`, what if it didn't allow you to write `f(x)`, it would always force line breaks after the `(` and before the `)`, meaning all your functions calls were forced to look like this:
```c++
    f(
        x
    )
```

See these parts of the test:

With a column limit, everything gets pulled onto one line, except if the template name is long:
```c++
  // now test that it handles the cases when the column limit forces wrapping
  Style.ColumnLimit = 40;
  // when the column limit allows it, the template should be combined back into
  // one line:
  verifyFormat("template <typename Foo, typename Bar>\n"
               "void foo() {}",
               "template <\n"
               "    typename Foo,\n"
               "    typename Bar\n"
               ">\n"
               "void foo() {}",
               Style);
  // but not when the name is looong
  verifyFormat("template <\n"
               "    typename Foo,\n"
               "    typename Barrrrrrrrrrrrrrrrrrrrrrrrrr\n"
               ">\n"
               "void foo() {}",
               Style);
...
  verifyFormat("void foo() { myFunc<T>(); }\n", Style);
  verifyFormat("void foo() {\n"
               "  myFunc<\n"
               "      Loooooooooooooooooooooooooooooooooooooooong\n"
               "  >();\n"
               "}\n",
               Style);
```

And for the no column limit case, these are the relevant tests to your question:

```c++
  verifyNoChange("template <typename Foo>\n"
                 "void foo() {}",
                 Style);
  verifyNoChange("template <typename Foo, typename Bar>\n"
                 "void foo() {}",
                 Style);
...
  // it should add a line break if not already present:
  verifyFormat("template <\n"
               "    typename Foo\n"
               ">\n"
               "void foo() {}",
               "template <\n"
               "    typename Foo>\n"
               "void foo() {}",
               Style);
  verifyFormat("template <\n"
               "    typename Foo,\n"
               "    typename Bar\n"
               ">\n"
               "void foo() {}",
               "template <\n"
               "    typename Foo,\n"
               "    typename Bar>\n"
               "void foo() {}",
               Style);
```

leijurv replied:
I believe the next test case shows this case:

```c++
      "  auto lambda =\n"
      "      [loooooooooooooooooooong]<\n"
      "          template <\n"
      "              typename Foooooooooooooooo,\n"
      "              typename\n"
      "          >\n"
      "          typename T\n"
      "      >(T t) {};\n"
      "}\n",
```

Because of the `Foooooooooooooooo`, it's unable to fit the `template <typename, typename>` onto one line, so it breaks both of them, and does indeed put the `>` onto the next line.

leijurv committed:
Merge branch 'main' into BreakBeforeTemplateClose

leijurv committed:
Merge branch 'main' into BreakBeforeTemplateClose

leijurv commented:
Ping :pray:

leijurv committed:
Merge branch 'main' into BreakBeforeTemplateClose

leijurv commented:
Thank you! It looks like I don't have permissions to merge, if you have them, could you?

owenca commented:
> > Maybe we should make it a sub-option for `BlockIndent`. I'm seeing a possible need for having control over other kinds of block indented structures such as Lambdas. I suspect the way this option has been implemented it interacts with the alignment options anyway. This would be a question probably for @owenca
> 
> I thought about it some more and I don't think it would be a good idea. Having a separate option seems a bit easier to handle, and allows mixing and matching by users.

I think this should be covered by [`BlockIndent`](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#alignafteropenbracket), which already covers both parentheses and braces. Let's hear what @mydeveloperday has to say.

owenca reviewed (commented):
About the test cases:
- Call `verifyFormat` instead of `verifyNoChange` whenever possible.
- Don't end a test case with a newline. (Search for `\n",` and `\n")`.)
- Capitalize the first word of a comment and end the comment with a `.` (or `:` if appropriate).

leijurv committed:
address comments

leijurv commented:
Thank you for the review @owenca 

> I think this should be covered by BlockIndent

My understanding from looking at past PRs to clang-format was that backwards compatibility was a top requirement. If I changed the behavior of BlockIndent to put the `>` on the next line, that would change a lot of existing behavior/codebases/tests.

> Call verifyFormat instead of verifyNoChange whenever possible.

I understand, but, I am testing the case where ColumnLimit = 0. When I call `verifyFormat` it removes all the newlines. So I was unable to test the behavior in the unconstrained column limit case. Note how I test that you can have a `\n` after the `template<`, or not, and this option doesn't force it either way.

In the later part of the tests, starting when I change the ColumnLimit to 40, note that all of the tests are `verifyFormat` and I don't use `verifyNoChange` anymore.

I have removed a few `verifyNoChange` in my last commit.

> Don't end a test case with a newline.
> Capitalize the first word of a comment and end the comment with a . (or : if appropriate).

Sure! Did so in this commit https://github.com/llvm/llvm-project/pull/118046/commits/19bc40e7d824804ebcd9839cb2155a704d368e36

owenca commented:
> > I think this should be covered by BlockIndent
> 
> My understanding from looking at past PRs to clang-format was that backwards compatibility was a top requirement. If I changed the behavior of BlockIndent to put the `>` on the next line, that would change a lot of existing behavior/codebases/tests.

The default behavior of `AlignAfterOpenBracket` is `Align` for `LLVM` style, and the breaking for template angles seems to have the same default behavior. Unfortunately, `WebKit` style has `DontAlign` as its default. So let's keep this new option separate.

> > Call verifyFormat instead of verifyNoChange whenever possible.
> 
> I understand, but, I am testing the case where ColumnLimit = 0. When I call `verifyFormat` it removes all the newlines. So I was unable to test the behavior in the unconstrained column limit case. Note how I test that you can have a `\n` after the `template<`, or not, and this option doesn't force it either way.
> 
> In the later part of the tests, starting when I change the ColumnLimit to 40, note that all of the tests are `verifyFormat` and I don't use `verifyNoChange` anymore.

I got that as I said "whenever possible".

> I have removed a few `verifyNoChange` in my last commit.

There are still a couple of `verifyNoChange` that can be replaced with `verifyFormat` for your `ColumnLimit = 0` test cases, though.

owenca commented on clang/include/clang/Format/Format.h:
```diff
@@ -2252,6 +2252,25 @@ struct FormatStyle {
+  /// If ``true``, a line break will be placed before the ``>`` in a multiline
+  /// template declaration.
+  /// \code
+  ///    true:
+  ///    template <
+  ///        typename Foo,
+  ///        typename Bar,
+  ///        typename Baz
+  ///    >
+  ///
+  ///    false:
+  ///    template <
+  ///        typename Foo,
+  ///        typename Bar,
+  ///        typename Baz>
```
We should make it an `enum` in order to handle more cases (now or in another patch), e.g.:
- Always
- BlockIndent
```
template <
    typename Foo, typename Bar,
    typename Baz
>
```
- Leave
- Multiline
```
template <typename Foo, typename Bar,
          typename Baz
>
```
- Never

leijurv replied:
In my mind that's a bit of a different option. You can see in your examples that you're assuming that the `>` goes on the next line, but that's what this PR is trying to add support for.

Like, your enum (which is totally reasonable) would switch between:

```
template <
    typename Foo, typename Bar,
    typename Baz
>
```

versus

```
template <typename Foo, typename Bar,
          typename Baz
>
```

But what I'm currently trying to add support for is:

```
template <
    typename Foo, typename Bar,
    typename Baz
>

template <typename Foo, typename Bar,
          typename Baz
>
```

versus

```
template <
    typename Foo, typename Bar,
    typename Baz>

template <typename Foo, typename Bar,
          typename Baz>
```

So I believe that should be another patch.

owenca replied:
> In my mind that's a bit of a different option. You can see in your examples that you're assuming that the `>` goes on the next line, but that's what this PR is trying to add support for.

`Always` would break before a template closer unconditionally.

> Like, your enum (which is totally reasonable) would switch between:
> 
> ```
> template <
>     typename Foo, typename Bar,
>     typename Baz
> >
> ```
> 
> versus
> 
> ```
> template <typename Foo, typename Bar,
>           typename Baz
> >
> ```
> 
> But what I'm currently trying to add support for is:
> 
> ```
> template <
>     typename Foo, typename Bar,
>     typename Baz
> >
> 
> template <typename Foo, typename Bar,
>           typename Baz
> >
> ```
> 
> versus
> 
> ```
> template <
>     typename Foo, typename Bar,
>     typename Baz>
> 
> template <typename Foo, typename Bar,
>           typename Baz>
> ```

I don't recall that you had a test case for the last pattern, but it seems that your current implementation is `Multiline`, which would break before the template closer if the template doesn't fit on one line regardless whether there is a break after the template opener (`BlockIndent`).

To clarify:
- BlockIndent
```
template <
    typename Foo, typename Bar,
    typename Baz
>

template <typename Foo, typename Bar,
          typename Baz>
```
- Multiline
```
template <
    typename Foo, typename Bar,
    typename Baz
>

template <typename Foo, typename Bar,
          typename Baz
>
```

> So I believe that should be another patch.

Like I said, I'm ok with some of the enumerated values being implemented in another patch.

leijurv replied:
Would `Always` look like:

```
template<typename T
>
```

For the overall indentation / line breaking of the various typenames, I left that untouched. All that stuff went way above my head. I didn't really have an intent in mind between your described options of BlockIndent vs Multiline but yes, I suppose what I have here is Multiline.

So, should I change this from a true/false to an enum of Multiline/Never?

owenca replied:
> Would `Always` look like:
> 
> ```
> template<typename T
> >
> ```

Yes, except that there is a space between `template` and `<` by `LLVM` default.

> So, should I change this from a true/false to an enum of Multiline/Never?

Yes. I haven't looked at your implementation yet, but are you doing `BreakBeforeTemplateCloser` or `TemplateCloserOnItsOwnLine`?

leijurv replied:
I'm doing `BreakBeforeTemplateCloser` because of [your comment here](https://github.com/llvm/llvm-project/pull/118046#discussion_r1912369946)

I've changed it to an enum in this commit: https://github.com/llvm/llvm-project/pull/118046/commits/9277a685c7a0199fff20677759d663254de9189f 👍

owenca replied:
I was asking whether you intended to also break _after_ the `>`, effectively forcing it to be on its own line.

leijurv replied:
No, I don't intend the `>` to be forced onto its own line. For instance, see this test case:

```
  verifyFormat("void foo() {\n"
               "  myFunc<\n"
               "      Loooooooooooooooooooooooooooooooooooooooong\n"
               "  >();\n"
               "}",
               Style);
```

owenca commented on clang/include/clang/Format/Format.h:
```diff
@@ -2252,6 +2252,25 @@ struct FormatStyle {
+  ///        typename Baz>
+  /// \endcode
+  /// \version 20
+  bool BreakBeforeTemplateClose;
```
```suggestion
  bool BreakBeforeTemplateCloser;
```

owenca commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -11116,6 +11116,275 @@ TEST_F(FormatTest, WrapsTemplateDeclarationsWithComments) {
+               "template <\n"
+               "    typename Foo,\n"
+               "    typename Bar\n"
+               ">"
```
```suggestion
               ">\n"
```

owenca reviewed (commented):
You can run `ninja clang-format-check-format` to make sure your patch is correctly formatted before pushing.

leijurv committed:
address comments

leijurv committed:
address comments

leijurv commented:
I apologize for forgetting `clang-format-check-format`

You're right that I had still missed some more cases of `verifyNoChange`. I've changed those now in https://github.com/llvm/llvm-project/pull/118046/commits/130428887e534cc3bfdfeec5846df2fd1e939667. The last two remaining `verifyNoChange` can't be changed to `verifyFormat`.

leijurv committed:
switch from bool to enum

owenca commented:
Why doesn't it break before the `>` for the example (a.cc) below?
```cpp
template <class A,
          class B>
class C {
  void f() {}
};
```
Command line:
```
clang-format -style='{BreakBeforeTemplateCloser: Multiline, ColumnLimit: 26}' a.cc
```

mydeveloperday reviewed (approved):
This looks good to me but please wait for @owenca to approve

leijurv committed:
switch from multiline to blockindent

leijurv commented:
Oh gosh, that's embarrassing. It looks like I was so focused on the column limit that I didn't cover cases where the column limit doesn't break the template fully. In other words, I would just type "Fooooooooooooooooooooooo" and force it onto the next line, without noticing that a slightly shorter Foooooo would have different behavior, I apologize.

In your example, I see that `BreakBeforeTemplateCloser: Multiline, ColumnLimit: 17` gives me:

```
template <
    class A,
    class B
>
class C {
  void f() {}
};
```

But `BreakBeforeTemplateCloser: Multiline, ColumnLimit: 18` gives me:

```
template <class A,
          class B>
class C {
  void f() {}
};
```

It looks to me like the problem is that when writing this code originally, I searched through the codebase for `FormatStyle::BAS_BlockIndent` and tried to replicate it. This causes the behavior here, where the `>` only goes onto the next line if there was a break after the matching `<`.

In other words, it seems **I was wrong** when I said `I didn't really have an intent in mind between your described options of BlockIndent vs Multiline but yes, I suppose what I have here is Multiline.` Actually, it seems like what I have here is BlockIndent.

So, thankfully I think I can fix this just by **renaming my option from Multiline to BlockIndent**, per your comment [here](https://github.com/llvm/llvm-project/pull/118046#discussion_r1912382882).

Additionally I have **added more tests** to cover this case where the first parameter can fit within the ColumnLimit on the same line as the `template <`.

https://github.com/llvm/llvm-project/pull/118046/commits/09528826ffc9f7b783d043c4447419cfd9a6ff38

leijurv committed:
one last test

leijurv committed:
rerun dump format style

leijurv committed:
Merge branch 'main' into BreakBeforeTemplateClose

leijurv committed:
Merge branch 'main' into BreakBeforeTemplateClose

owenca commented on clang/lib/Format/ContinuationIndenter.cpp:
```diff
@@ -406,6 +406,10 @@ bool ContinuationIndenter::mustBreak(const LineState &State) {
   }
   if (CurrentState.BreakBeforeClosingParen && Current.is(tok::r_paren))
     return true;
+  if (CurrentState.BreakBeforeClosingAngle && Current.is(TT_TemplateCloser) &&
+      Style.BreakBeforeTemplateCloser == FormatStyle::BBTCS_BlockIndent) {
+    return true;
+  }
```
```suggestion
  if (CurrentState.BreakBeforeClosingAngle && Current.is(TT_TemplateCloser))
    return true;
```
See below.

owenca commented on clang/lib/Format/ContinuationIndenter.cpp:
```diff
@@ -1235,6 +1239,9 @@ unsigned ContinuationIndenter::addTokenOnNewLine(LineState &State,
         Style.AlignAfterOpenBracket == FormatStyle::BAS_BlockIndent;
   }
 
+  if (PreviousNonComment && PreviousNonComment->is(tok::less))
+    CurrentState.BreakBeforeClosingAngle = true;
```
```suggestion
  if (PreviousNonComment && PreviousNonComment->is(TT_TemplateOpener)) {
    CurrentState.BreakBeforeClosingAngle =
        Style.BreakBeforeTemplateCloser == FormatStyle::BBTCS_BlockIndent;
  }
```

owenca commented on clang/lib/Format/ContinuationIndenter.cpp:
```diff
@@ -1371,6 +1378,11 @@ unsigned ContinuationIndenter::getNewLineColumn(const LineState &State) {
       State.Stack.size() > 1) {
     return State.Stack[State.Stack.size() - 2].LastSpace;
   }
+  if (Current.is(TT_TemplateCloser) &&
+      Style.BreakBeforeTemplateCloser != FormatStyle::BBTCS_Never &&
+      State.Stack.size() > 1) {
```
```suggestion
  if (Style.BreakBeforeTemplateCloser == FormatStyle::BBTCS_BlockIndent &&
      Current.is(TT_TemplateCloser) && State.Stack.size() > 1) {
```

leijurv committed:
Merge branch 'main' into BreakBeforeTemplateClose

leijurv committed:
address comments

leijurv commented:
Thank you for the review owenca, those are very reasonable changes, particularly on the conditions where the BreakBeforeClosingAngle gets set/used!

owenca commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -11220,6 +11220,333 @@ TEST_F(FormatTest, WrapsTemplateDeclarationsWithComments) {
+               ">\n"
+               "void foo() {}",
+               Style);
+  // But not when the name is looong. Note that these names are exactly 1
+  // character too long for the ColumnLimit.
+  verifyFormat("template <\n"
+               "    typename Foo,\n"
+               "    typename Barrrrrrrrrrrrrrrrrrrr\n"
+               ">\n"
+               "void foo() {}",
+               Style);
```
Is this just 1 over `ColumnLimit`? Looks like we got a problem here:
```
$ cat a.cc
template <typename Barrrrrrrrrrrrrrrrrrr>
void foo() {}
$ clang-format -style='{ColumnLimit: 40, BreakBeforeTemplateCloser: BlockIndent}' a.cc
template <typename Barrrrrrrrrrrrrrrrrrr
>
void foo() {}
$ 
```
Shouldn't the output be the following instead?
```cpp
template <
    typename Barrrrrrrrrrrrrrrrrrr
>
void foo() {}
```
In comparison, both 19.1.7 and the default `Never` of this patch give the expected output below:
```
$ clang-format -style='{ColumnLimit: 40}' a.cc
template <
    typename Barrrrrrrrrrrrrrrrrrr>
void foo() {}
$ 
```

leijurv replied:
I apologize for the off-by-one. You're correct that you've found unexpected behavior - when the column limit is 1 less than `template <typename Barr` then it splits properly, when the column limit is 1 more, then it allows it all on one line, but when the column limit is exactly equal to the length, it chooses to split before the `<` when we'd rather it split after the `<`.

I'm not certain whether to call this behavior "wrong", but sure it is unexpected. I would like to point out that I had already encountered this behavior with the lambdas, and I had written it into the unit test:
https://github.com/llvm/llvm-project/blob/fa15aca5c58a16760479fa3e53fa1286fc6adffa/clang/unittests/Format/FormatTest.cpp#L11455-L11468 I was okay with that behavior however it contradicts your suggestion here so I have trimmed down and modified the tests because I don't want to lock in behavior that is unexpected. I tried to make the lambda behavior make sense alongside this change, but, between the two, it is clear that lambda templates are not the common case.

See commit: https://github.com/llvm/llvm-project/pull/118046/commits/c0090c8ae6bea4091bc061601902f99903c1a89f

leijurv replied:
I also made sure your example case is fixed, and added a unit test accordingly too:
```
$ ./bin/clang-format -style='{ColumnLimit: 41, BreakBeforeTemplateCloser: BlockIndent}' chk.cpp 
template <typename Barrrrrrrrrrrrrrrrrrr>
void foo() {}
$ ./bin/clang-format -style='{ColumnLimit: 40, BreakBeforeTemplateCloser: BlockIndent}' chk.cpp
template <
    typename Barrrrrrrrrrrrrrrrrrr
>
void foo() {}
```

leijurv replied:
Let me clarify my earlier comment.

For this PR, I don't feel as though it's on me to ensure clang-format consistently / aesthetically picks between these:

```c++
// good:
template <typename Foo, typename Bar>

template <typename Foo,
          typename Bar>

template <
   typename Foo,
   typename Bar
>
```

The only thing that I need to do in this PR is prevent these cases, which should not be allowed when my feature is turned on (per your most recent suggestion, which I'm willing to accept):

```c++
// bad:
template <typename Foo, typename Bar
>

template <typename Foo,
          typename Bar
>

template <
   typename Foo,
   typename Bar>
```

As long as the formatter is picking between one those first three, and never one of the latter three, I think we're in the clear.

The issue arose last night when I was updating the lambda tests. It was picking between one of the first three, but, not consistently. It depended on whether "template Loooong" or "template T" was first or second. That's silly behavior that I don't agree with. So I intentionally did not test it. All I'm testing is that it picks between one of the allowable three variants, within reason. I do not intend or desire to lock in exactly which lambda scenarios cause which to be picked, because that isn't what I'm intending to change in this PR. So if it's inconsistent between two options, both of which are reasonable, I just decided to remove that from my tests.

gedare replied:
I just fixed a similar bug in `BlockIndent`, see https://github.com/llvm/llvm-project/pull/124998

leijurv replied:
Great point gedare, in order to fix this issue I first tried to find where BlockIndent did the same thing - except I couldn't find it. Then I tried to fix it by adding a penalty - also didn't work. It took a while before I eventually found `ContinuationIndenter::canBreak`, which appears to me to be the correct solution, which you are also using as well in that PR 👍 glad that we agree

owenca replied:
> Let me clarify my earlier comment.
> 
> For this PR, I don't feel as though it's on me to ensure clang-format consistently / aesthetically picks between these:
> 
> ```c++
> // good:
> template <typename Foo, typename Bar>
> 
> template <typename Foo,
>           typename Bar>
> 
> template <
>    typename Foo,
>    typename Bar
> >
> ```
> 
> The only thing that I need to do in this PR is prevent these cases, which should not be allowed when my feature is turned on (per your most recent suggestion, which I'm willing to accept):
> 
> ```c++
> // bad:
> template <typename Foo, typename Bar
> >
> 
> template <typename Foo,
>           typename Bar
> >
> 
> template <
>    typename Foo,
>    typename Bar>
> ```
> 
> As long as the formatter is picking between one those first three, and never one of the latter three, I think we're in the clear.
> 
> The issue arose last night when I was updating the lambda tests. It was picking between one of the first three, but, not consistently. It depended on whether "template Loooong" or "template T" was first or second. That's silly behavior that I don't agree with. So I intentionally did not test it. All I'm testing is that it picks between one of the allowable three variants, within reason. I do not intend or desire to lock in exactly which lambda scenarios cause which to be picked, because that isn't what I'm intending to change in this PR. So if it's inconsistent between two options, both of which are reasonable, I just decided to remove that from my tests.

If I understand you correctly, you just want to add a boolean option to break after the template closer when there is a break after the matching template opener? I'm fine with that! It seems though making it a struct doesn't make sense anymore. Can we change it back to a boolean?

leijurv replied:
Hm. Not really what I meant - I don't mind if it's a boolean or an enum, I guess slight preference for enum because I could imagine people have other opinions on when to break, similar to AlignAfterOpenBracket. So I didn't mean to say that I prefer a boolean.

All I meant to say is that the specific code that decides which of these happens:

```
template <typename Foo,
          typename Bar>
```

versus

```
template <
   typename Foo,
   typename Bar
>
```

was being inconsistent/confusing, and, I didn't want to make it consistent in this PR. Importantly, I feel this way because this PR is not intended to affect that behavior. I only intend to affect where the `>` goes in that second case. Not whether the first or second case is selected.

I really just meant to explain why I had removed some tests that were previously there.

I'm fine with this PR as-is, but if these last few messages have changed something for you, I'll change it back to boolean if you say so, but my preference (currently) is enum.

owenca replied:
`AlignAfterOpenBracket` is for breaking after the opening bracket; only its `BlockIndent` is for whether to also break after the closing bracket. It makes more sense to me to make this patch either about breaking after the template opener (with `Always`, `BlockIndent`, etc.) or about breaking before the template closer if there's a break after the matching template opener (a boolean).

leijurv replied:
Okay, understood. I'll change it back to a boolean then.

leijurv replied:
All right @owenca how does this look: https://github.com/llvm/llvm-project/pull/118046/commits/a768adb5a38ddad8bf1add7686d7681dac7ece81

leijurv committed:
address comments

leijurv committed:
Merge branch 'main' into BreakBeforeTemplateClose

leijurv committed:
Merge branch 'main' into BreakBeforeTemplateClose

leijurv committed:
Merge branch 'main' into BreakBeforeTemplateClose

leijurv committed:
bump version to 21

leijurv committed:
also need to bump version here

leijurv committed:
switch back to bool

owenca commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -11224,6 +11224,328 @@ TEST_F(FormatTest, WrapsTemplateDeclarationsWithComments) {
+  // character too long for the ColumnLimit.
+  verifyFormat("template <\n"
+               "    typename Foo,\n"
+               "    typename Barrrrrrrrrrrrrrrrrrrr\n"
```
```suggestion
               "    typename Barrrrrrrrrrrrrrrrrrr\n"
```

owenca commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -11224,6 +11224,328 @@ TEST_F(FormatTest, WrapsTemplateDeclarationsWithComments) {
+  verifyFormat(
+      "template <\n"
+      "    typename Foo,\n"
+      "    typename Barrrrrrrrrrrrrrrrrrrrrrrrrr\n"
```
```suggestion
      "    typename Barrrrrrrrrrrrrrrrrrr\n"
```

owenca commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -11224,6 +11224,328 @@ TEST_F(FormatTest, WrapsTemplateDeclarationsWithComments) {
+      "    typename Barrrrrrrrrrrrrrrrrrrrrrrrrr\n"
+      ">\n"
+      "void foo() {}",
+      "template <typename Foo, typename Barrrrrrrrrrrrrrrrrrrrrrrrrr>\n"
```
```suggestion
      "template <typename Foo, typename Barrrrrrrrrrrrrrrrrrr>\n"
```

owenca commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -11224,6 +11224,328 @@ TEST_F(FormatTest, WrapsTemplateDeclarationsWithComments) {
+      Style);
+  verifyFormat(
+      "template <\n"
+      "    typename Fooooooooooooooooooooooooooo,\n"
```
```suggestion
      "    typename Foooooooooooooooooooo,\n"
```

owenca commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -11224,6 +11224,328 @@ TEST_F(FormatTest, WrapsTemplateDeclarationsWithComments) {
+      "    typename Bar\n"
+      ">\n"
+      "void foo() {}",
+      "template <typename Fooooooooooooooooooooooooooo, typename Bar>\n"
```
```suggestion
      "template <typename Foooooooooooooooooooo, typename Bar>\n"
```

owenca commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -11224,6 +11224,328 @@ TEST_F(FormatTest, WrapsTemplateDeclarationsWithComments) {
+      Style);
+  // Even when there is only one long name:
+  verifyFormat("template <\n"
+               "    typename Fooooooooooooooooooooooooooo\n"
```
```suggestion
               "    typename Foooooooooooooooooooo\n"
```

owenca commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -11224,6 +11224,328 @@ TEST_F(FormatTest, WrapsTemplateDeclarationsWithComments) {
+               "    typename Fooooooooooooooooooooooooooo\n"
+               ">\n"
+               "void foo() {}",
+               "template <typename Fooooooooooooooooooooooooooo>\n"
```
```suggestion
               "template <typename Foooooooooooooooooooo>\n"
```

owenca commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -11224,6 +11224,328 @@ TEST_F(FormatTest, WrapsTemplateDeclarationsWithComments) {
+  verifyFormat("void foo() { myFunc<T>(); }", Style);
+  verifyFormat("void foo() {\n"
+               "  myFunc<\n"
+               "      Loooooooooooooooooooooooooooooooooooooooong\n"
```
```suggestion
               "      Looooooooooooooooooooooooong\n"
```

leijurv commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -11224,6 +11224,328 @@ TEST_F(FormatTest, WrapsTemplateDeclarationsWithComments) {
+               Style);
+
+  Style.BreakBeforeTemplateCloser = true;
+  // BreakBeforeTemplateCloser should NOT force template declarations onto
+  // multiple lines.
+  verifyFormat("template <typename Foo>\n"
+               "void foo() {}",
+               Style);
+  verifyFormat("template <typename Foo, typename Bar>\n"
+               "void foo() {}",
+               Style);
```
Okay deleted, but why is this not a reasonable test? The point is that turning on BreakBeforeTemplateCloser does NOT change the current behavior.

owenca replied:
You can add them back if you prefer.

leijurv replied:
I apologize, it appears that in my commit https://github.com/leijurv/llvm-project/commit/e1eaba9b05b00adf44628e2ceb1b3374fec902cb I missed removing this one, bit of a coincidence (I swear it wasn't on purpose). I have just now triple checked to ensure that all your other comments were indeed addressed, I believe I only missed this one.

leijurv committed:
address comments

leijurv commented:
Thank you so much for taking the time to review the tests! I apologize for all the redundancies.

I believe I have addressed all of your comments here: https://github.com/leijurv/llvm-project/commit/e1eaba9b05b00adf44628e2ceb1b3374fec902cb

leijurv committed:
empty commit to kick github

owenca commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -11224,6 +11224,328 @@ TEST_F(FormatTest, WrapsTemplateDeclarationsWithComments) {
+  verifyFormat("template <typename Foo, typename Bar>\n"
+               "void foo() {}",
+               Style);
+  // It should allow a line break, even when the typename is short.
+  // verifyNoChange is needed because the default behavior is one line.
+  verifyNoChange("template <\n"
+                 "    typename Foo\n"
+                 ">\n"
+                 "void foo() {}",
+                 Style);
+  verifyNoChange("template <\n"
+                 "    typename Foo,\n"
+                 "    typename Bar\n"
+                 ">\n"
+                 "void foo() {}",
+                 Style);
```
As I explained, they are already tested by "the test cases at lines 11272 and 11280 below" as the two-argument version of `verifyFormat()` first checks that the input is formatted to the expected output and then ensures that the expected output doesn't change when being formatted. In other words, `verifyFormat(Output, Input)` also does `verifyNoChange(Output)`.

leijurv replied:
I was unaware of that! That explains a lot of redundancies that I now agree with you about. Removed: https://github.com/leijurv/llvm-project/commit/b6219e250b80114c9b93a7762db7c3e0cbd84260

leijurv commented:
GitHub is having issues it appears - the commit https://github.com/leijurv/llvm-project/commit/e1eaba9b05b00adf44628e2ceb1b3374fec902cb is certainly pushed, and I even tried pushing an empty commit on top to kick github into working, https://github.com/leijurv/llvm-project/commit/7d456ca2457dc7d7d2403562386097880bedaa34 but still this PR is in a glitched state that doesn't show those commits...

<img width="330" alt="Screen Shot 2025-02-04 at 7 35 58 PM" src="https://github.com/user-attachments/assets/9b48ee20-9da8-434c-8986-9445f143f2b9" />

leijurv committed:
address comment

leijurv commented:
Ah, it resolved itself as of https://github.com/leijurv/llvm-project/commit/b6219e250b80114c9b93a7762db7c3e0cbd84260

owenca commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -11224,6 +11224,328 @@ TEST_F(FormatTest, WrapsTemplateDeclarationsWithComments) {
+               "  auto lambda = []<typename T>(T t) {};\n"
+               "}",
+               Style);
+
```
That suggestion was aimed at shortening the test case so that we don't need to break the string in the middle of the line. I wasn't thinking about redundant test cases then.

leijurv replied:
Gotcha

owenca commented on clang/lib/Format/ContinuationIndenter.cpp:
```diff
@@ -304,6 +304,10 @@ bool ContinuationIndenter::canBreak(const LineState &State) {
                                    Current.closesBlockOrBlockTypeList(Style))) {
     return false;
   }
+  if (Style.BreakBeforeTemplateCloser && Current.is(TT_TemplateCloser) &&
+      !CurrentState.BreakBeforeClosingAngle) {
+    return false;
+  }
```
```suggestion
  if (Style.BreakBeforeTemplateCloser && Current.is(TT_TemplateCloser))
    return CurrentState.BreakBeforeClosingAngle;
```
And move this to line 362 below for ease of maintenance.

leijurv commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -11224,6 +11224,254 @@ TEST_F(FormatTest, WrapsTemplateDeclarationsWithComments) {
+
+  // Now test that it handles the cases when the column limit forces wrapping.
+  Style.ColumnLimit = 40;
+  // Breaks when the name is looong. Note that these names are exactly 1
+  // character too long for the ColumnLimit.
+  verifyFormat("template <\n"
+               "    typename Foo,\n"
+               "    typename Barrrrrrrrrrrrrrrrrrr\n"
+               ">\n"
+               "void foo() {}",
+               Style);
+  // BlockIndent style is used when the ColumnLimit allows it:
+  verifyFormat("template <typename Foo,\n"
+               "          typename Barrrrrrrrrrrrrrrrrr>\n"
+               "void foo() {}",
+               Style);
+  verifyFormat("template <typename Fooooooooooooooooooo,\n"
+               "          typename Bar>\n"
+               "void foo() {}",
+               Style);
```
Okay deleted

I can appreciate that these two were redundant:
```c++
  verifyFormat("template <\n"
               "    typename Foo,\n"
               "    typename Barrrrrrrrrrrrrrrrrrr\n"
               ">\n"
               "void foo() {}",
               Style);
  verifyFormat("template <\n"
               "    typename Foo,\n"
               "    typename Barrrrrrrrrrrrrrrrrrr\n"
               ">\n"
               "void foo() {}",
               "template <typename Foo, typename Barrrrrrrrrrrrrrrrrrr>\n"
               "void foo() {}",
               Style);
```

However, these other two I don't think are tested anymore?
```c++
  verifyFormat("template <typename Foo,\n"
               "          typename Barrrrrrrrrrrrrrrrrr>\n"
               "void foo() {}",
               Style);
  verifyFormat("template <typename Fooooooooooooooooooo,\n"
               "          typename Bar>\n"
               "void foo() {}",
               Style);
```

I can appreciate that it's current behavior, so, it's not that important to test that it's unchanged with the feature on.

leijurv replied:
https://github.com/llvm/llvm-project/pull/118046/commits/4faf28a1b3752d2ddbf88f41b74d11dc9290aefc

leijurv committed:
address comments

owenca commented on clang/include/clang/Format/Format.h:
```diff
@@ -2252,6 +2252,33 @@ struct FormatStyle {
   /// \version 16
   BreakBeforeInlineASMColonStyle BreakBeforeInlineASMColon;
 
+  /// If ``true``, a line break will be placed before the ``>`` in a block
+  /// indented template declaration.
```
```suggestion
  /// If ``true``, break before a template closing bracket (``>``) when there is
  /// a line break after the matching opening bracket (``<``).
```

leijurv committed:
address comments

leijurv committed:
Update clang/include/clang/Format/Format.h

Co-authored-by: Owen Pan <owenpiano@gmail.com>

leijurv committed:
regenerate docs

owenca committed:
NFC: minor cleanup

owenca committed:
Remove trailing whitespace in ContinuationIndenter.cpp

lengye7 commented:
Can this option be used to format any template like std::remove_cv_t<add_common_cv_reference<std::common_type_t<std::decay_t<T0>, std::decay_t<T1>>,T0,T1>>;?

leijurv commented:
> Can this option be used to format any template like std::remove_cv_t<add_common_cv_reference<std::common_type_t<std::decay_t, std::decay_t>,T0,T1>>;?

Yes! In fact, that exact case is tested.

https://github.com/llvm/llvm-project/commit/d2b45ce100d641a8f1690e30843bb9c5ea71ab86#diff-3f6f57cda9809a57c5b79e22b4181b3f3aaac7216262d0ef44108f39b0443e9bR11276-R11290

```c++
  // Test from https://github.com/llvm/llvm-project/issues/80049:
  verifyFormat(
      "using type = std::remove_cv_t<\n"
      "    add_common_cv_reference<\n"
      "        std::common_type_t<std::decay_t<T0>, std::decay_t<T1>>,\n"
      "        T0,\n"
      "        T1\n"
      "    >\n"
      ">;",
      "using type = std::remove_cv_t<\n"
      "    add_common_cv_reference<\n"
      "        std::common_type_t<std::decay_t<T0>, std::decay_t<T1>>,\n"
      "        T0,\n"
      "        T1>>;",
      Style);
```
